File: exercicios/projeto-final/skeletons/thecocktaildb/flutter/pratica/lib/screens/list_screen.dart
import 'package:flutter/material.dart';
import '../api/cocktaildb_api.dart';
import '../models/drink.dart';
import 'detail_screen.dart';

const List<String> _categories = ['shot', 'beer', 'shake'];

String _apiCategory(String slug) {
  switch (slug) {
    case 'shot':
      return 'Shot';
    case 'beer':
      return 'Beer';
    case 'shake':
      return 'Shake';
    default:
      return 'Cocktail';
  }
}

class ListScreen extends StatefulWidget {
  const ListScreen({super.key});

  @override
  State<ListScreen> createState() => _ListScreenState();
}

class _ListScreenState extends State<ListScreen> {
  final CocktailDbApi _api = CocktailDbApi();

  List<DrinkSummary> _all = [];
  bool _loading = true;
  String? _error;
  String _searchText = '';
  String? _selectedCategory;
  Set<String>? _categoryNames;

  @override
  void initState() {
    super.initState();
    _loadList();
  }

  Future<void> _loadList() async {
    try {
      final result = await _api.fetchList();
      setState(() {
        _all = result;
        _loading = false;
      });
    } catch (error) {
      print(error.toString());
      setState(() {
        _error = error.toString();
        _loading = false;
      });
    }
  }

  Future<void> _selectCategory(String? category) async {
    try {
      if (category == null) {
        setState(() {
          _selectedCategory = null;
          _categoryNames = null;
        });
      } else {
        final result = await _api.fetchNamesByCategory(_apiCategory(category));
        setState(() {
          _categoryNames = result.toSet();
          _selectedCategory = category;
        });
      }
    } catch (error) {
      print(error.toString());
      setState(() {
        _error = error.toString();
        _loading = false;
      });
    }
  }

  List<DrinkSummary> get _filtered {
    // TODO 3 (feature 3 — busca) + TODO 4 (feature 4 — categoria): filtrar
    // `_all` por `_categoryNames` (quando não-nulo, `_categoryNames!.contains`)
    // e por `_searchText` (substring case-insensitive do `name`).
    return _all.where((drink) {
          final matchesCategory =
              _categoryNames == null || _categoryNames!.contains(drink.name);
          final matchesSearch =
              drink.name.toLowerCase().contains(_searchText.toLowerCase());
          return matchesCategory && matchesSearch;
        }).toList();
      }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Drinks')),
      body: Semantics(identifier: 'item-list-screen', child: _buildBody()),
    );
  }

  Widget _buildBody() {
    if (_loading) {
      return const Center(child: CircularProgressIndicator());
    }
    if (_error != null) {
      return Center(child: Text(_error!));
    }
    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.all(12),
          child: Semantics(
            identifier: 'search-input',
            child: TextField(
              decoration: const InputDecoration(
                labelText: 'Buscar drink',
                border: OutlineInputBorder(),
                prefixIcon: Icon(Icons.search),
              ),
              onChanged: (value) => setState(() => _searchText = value),
            ),
          ),
        ),
        SizedBox(
          height: 44,
          child: ListView(
            scrollDirection: Axis.horizontal,
            padding: const EdgeInsets.symmetric(horizontal: 12),
            children: [
              _buildCategoryChip(null, 'Todos'),
              for (final category in _categories)
                _buildCategoryChip(category, category),
            ],
          ),
        ),
        const SizedBox(height: 8),
        Expanded(
          child: ListView.builder(
            itemCount: _filtered.length,
            itemBuilder: (context, index) {
              final drink = _filtered[index];
              return Semantics(
                identifier: 'item-card-${drink.id}',
                child: ListTile(
                  leading: Image.network(drink.thumbUrl, width: 40, height: 40),
                  title: Text(drink.name),
                  onTap: () {
                    Navigator.of(context).push(
                      MaterialPageRoute(
                        builder: (_) => DetailScreen(drinkId: drink.id),
                      ),
                    );
                  },
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _buildCategoryChip(String? category, String label) {
    final selected = _selectedCategory == category;
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 4),
      child: Semantics(
        identifier: category == null
            ? 'category-chip-all'
            : 'category-chip-$category',
        child: ChoiceChip(
          label: Text(capitalize(label)),
          selected: selected,
          onSelected: (_) => _selectCategory(category),
        ),
      ),
    );
  }
}

File: exercicios/projeto-final/skeletons/thecocktaildb/kmp/pratica/androidApp/src/main/kotlin/com/puciec/cocktailkmp/ListScreen.kt
package com.puciec.cocktailkmp

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.puciec.cocktailkmp.data.TheCocktailDbApi
import com.puciec.cocktailkmp.data.DrinkSummary
import com.puciec.cocktailkmp.data.capitalize

private val CATEGORIES = listOf("shot", "beer", "shake")

private fun apiCategory(slug: String): String = when (slug) {
    "shot" -> "Shot"
    "beer" -> "Beer"
    "shake" -> "Shake"
    else -> "Cocktail"
}

@Composable
fun ListScreen(api: TheCocktailDbApi, onSelect: (String) -> Unit) {
    var all by remember { mutableStateOf<List<DrinkSummary>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }
    var searchText by remember { mutableStateOf("") }
    var selectedCategory by remember { mutableStateOf<String?>(null) }
    var categoryNames by remember { mutableStateOf<Set<String>?>(null) }

    LaunchedEffect(Unit) {
        try {
            all = api.fetchList()
        } catch (e: Exception) {
            error = e.message
        } finally {
            loading = false
        }
    }

    val filtered = all.filter { it.strDrink.contains(searchText, ignoreCase = true) }

    if (loading) {
        Box(
            Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) { CircularProgressIndicator() }
        return
    }
    if (error != null) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { Text(error!!) }
        return
    }

    Box(
        Modifier
            .fillMaxSize()
            .testTag("item-list-screen")
    ) {
        LazyColumn(Modifier.fillMaxSize()) {
            item {
                OutlinedTextField(
                    value = searchText,
                    onValueChange = { searchText = it },
                    label = { Text("Buscar drink") },
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(12.dp)
                        .testTag("search-input"),
                )
            }
            item {
                LazyRow(
                    Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 12.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    item {
                        FilterChip(
                            selected = selectedCategory == null,
                            onClick = { selectedCategory = null; categoryNames = null },
                            label = { Text("Todos") },
                            modifier = Modifier.testTag("category-chip-all"),
                        )
                    }
                    items(CATEGORIES) { category ->
                        FilterChip(
                            selected = selectedCategory == category,
                            onClick = { selectedCategory = category },
                            label = { Text(capitalize(category)) },
                            modifier = Modifier.testTag("category-chip-$category"),
                        )
                    }
                }
            }
            items(filtered, key = { it.idDrink }) { drink ->
                Text(
                    text = drink.strDrink,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag("item-card-${drink.idDrink}")
                        .clickable { onSelect(drink.idDrink) }
                        .padding(16.dp),
                    style = MaterialTheme.typography.bodyLarge,
                )
            }
        }
    }

    LaunchedEffect(selectedCategory) {
        try {
            all = if (selectedCategory == null) {
                api.fetchList()
            } else {
                api.fetchList(apiCategory(selectedCategory!!))
            }
        } catch (e: Exception) {
            error = e.message
        }
    }
}
